### ISSUE TITLE
[test_scan2d]: Complete support for `cumsum` and `cumprod`

### ISSUE THREAD

**Issue by etiotto** [intel/intel-xpu-backend-for-triton#153]
The end-to-end core test `test_scan2d` fails to run correctly on a PVC GPU for certain combinations of arguments/attributes (axis = 1, num_warps = 4, shape = [1024, 2]).

**Comment by ienkovich** [comment]
I looked into LLVM IR dumps for the case with `op=cumsum, shape = [16, 32], axis=0, num_warps = 4`. Each thread handles four elements in a row. Each warp has 32 threads and handles four input rows. Here is a code to compute the partial sum for the first element handled by a thread:
```
  %26 = tail call i32 @_Z21sub_group_shuffle_xorij(i32 %19, i32 8), !dbg !14
  %27 = icmp eq i32 %9, 0, !dbg !14
  %28 = select i1 %27, i32 0, i32 %26, !dbg !14
  %29 = add i32 %28, %19, !dbg !14
  %30 = tail call i32 @_Z21sub_group_shuffle_xorij(i32 %29, i32 16), !dbg !14
  %31 = icmp ult i32 %9, 2, !dbg !14
  %32 = select i1 %31, i32 0, i32 %30, !dbg !14
  %33 = add i32 %29, %32, !dbg !14
```
Here `%19` is an input value and `%9` is a row index within a warp. More readable C code for this:
```
other_val = sub_group_shuffle_xor(val, 8)
local_sum = val + row_no== 0 ? 0 : other_val
other_local_sum = sub_group_shuffle_xor(local_sum, 16)
res = local_sum + row_no < 2 ? 0 : other_local_sum
```
We expect `res` to hold warp-local partial sums in each thread. But the condition (`row_no == 0`) applied to the `local_sum` allows adding elements from subsequent rows into the partial sum, so row#3 is added to row#2. Also, the second sub-group shuffle exchanges partial sums instead of the full sum for the appropriate half of a warp.

E.g. here is what happens now for the first column within the first warp:
```
row_no    val      other_val       local_sum         other_local_sum          sum
    0    x[0, 0]    x[1, 0]    x[0, 0]              x[2, 0] + x[3, 0]      x[0, 0]
    1    x[1, 0]    x[0, 0]    x[0, 0] + x[1, 0]    x[2, 0] + x[3, 0]      x[0, 0] + x[1, 0]
    2    x[2, 0]    x[3, 0]   *x[2, 0] + x[3, 0]*  *x[0, 0]*              *x[0, 0] + x[2, 0] + x[3, 0]*
    3    x[3, 0]    x[2. 0]    x[2, 0] + x[3, 0]    x[0, 0] + x[1, 0]      x[0, 0] + x[1, 0] + x[2, 0] + x[3, 0]
```
Here we see that row#2 wrongly includes row#3 in its sum and misses row#1. The same thing happens for all columns and each warp ends up with one incorrect row. 

To fix that, we should fix the condition used for partial sum computation (which is now `row_no == 0`) and at each reduction stage compute full sums for exchange via `sub_group_shuffle_xor`. At this moment, I don't know which part of the lowering is responsible for the error.

**Comment by ienkovich** [comment]
The root cause of the problem is in `commonShflSync` method which ignores shuffle mode for GenX target.

**Comment by ienkovich** [comment]
Fixed by #218
